File: java/BOJ2156.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ2156 {
    // https://www.acmicpc.net/problem/

    // 변수 설정
    static FastReader fr = new FastReader();
    static int n;
    static int[] wine;
    static int[][] Dy;

    // 입력 함수
    static void input(){
        n = fr.nextInt();
        Dy = new int[n+1][3];
        wine = new int[n+1];

        for (int i = 1; i <= n; i++) {
            wine[i] = fr.nextInt();
        }
    }

    static void dp(){
        int answer = 0;
        Dy[1][0] = wine[1];
        Dy[1][1] = wine[1];
        Dy[1][2] = 0;

        for (int i = 2; i <= n; i++) {
            Dy[i][2] = answer;
            Dy[i][0] = Math.max(Math.max(Dy[i-2][0], Dy[i-2][1]), Dy[i-1][2]) + wine[i];
            Dy[i][1] = Dy[i-1][0] + wine[i];
            answer = Math.max(Dy[i][0], answer);
            answer = Math.max(Dy[i][1], answer);
        }

        System.out.println(n == 1? wine[1] : answer);
    }

    public static void main(String[] args) throws Exception {
        input();
        dp();
    }

    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: py/BOJ/BOJ2156.py
import sys

input = sys.stdin.readline
# https://www.acmicpc.net/problem/2156

n = int(input())
wines = [0]

for _ in range(n):
    wines.append(int(input()))

"""
dy[i][0~2] : i번째 와인을 마실 때, 이전 와인을 0잔 ~ 2잔 마신경우
dy[i][0] : i - 2번째 위치의 최대값 + 현재 와인 양
dy[i][1] : i - 1번째 위치의 이전에 아무것도 안마신 양 + 현재 와인 양
dy[i][2] : i - 1번째 위치의 이전 한잔도 마신 양(현재 와인은 못마심)
"""
dy = [[0] * 3, [wines[1]] * 3]

for i in range(2, n + 1):
    dy.append([max(dy[i - 2]) + wines[i], dy[i - 1][0] + wines[i], dy[i - 1][1]])

print(max(dy[-1]))
